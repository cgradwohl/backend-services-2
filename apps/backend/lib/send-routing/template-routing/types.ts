import { IRoutingSummary } from "~/send/worker/commands/route/types";

export type TemplateV1RouteNode = IRoutingSummary & {
  canUseForFailover?: boolean;
};

export type TemplateV1RoutingSummary = {
  always: TemplateV1RouteNode[];
  bestOf: TemplateV1RouteNode[];
};
