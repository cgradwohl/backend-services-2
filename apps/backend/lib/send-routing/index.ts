export * from "./routing-strategy-store";
export * from "./types";
export { generateRouting } from "./generate-routing";
export * from "./timeout-helpers";
export { getAllRouteLeafs, GetAllRouteLeafsOpts, getRouteNode } from "./lib";
export * from "./get-failover-node";
export * from "./template-routing";
export * from "./errors";
