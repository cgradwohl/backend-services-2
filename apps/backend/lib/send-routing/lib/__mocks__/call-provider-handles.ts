const { callProviderHandles: actualCallProviderHandles } = jest.requireActual(
  "~/lib/send-routing/lib/call-provider-handles"
);
export const callProviderHandles = jest.fn(actualCallProviderHandles);
