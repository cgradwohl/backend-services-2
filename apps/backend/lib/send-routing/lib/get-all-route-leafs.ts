import { DeadRouteBranch, DeadRouteLeaf, RouteLeaf, RouteNode } from "../types";

export interface GetAllRouteLeafsOpts {
  includeFailoverBranches?: boolean;
  includeDeadLeafs?: boolean;
}

export interface GetAllRouteLeafsWithDeadLeafsOpts
  extends GetAllRouteLeafsOpts {
  includeDeadLeafs: true;
}

export function getAllRouteLeafs(
  tree: RouteNode,
  opts: GetAllRouteLeafsWithDeadLeafsOpts
): (RouteLeaf | DeadRouteLeaf | DeadRouteBranch)[];
export function getAllRouteLeafs(
  tree: RouteNode,
  opts?: GetAllRouteLeafsOpts
): RouteLeaf[];
export function getAllRouteLeafs(
  tree: RouteNode,
  opts?: GetAllRouteLeafsOpts
): (RouteLeaf | DeadRouteLeaf | DeadRouteBranch)[] {
  if (tree.type === "branch") {
    return [
      ...tree.nodes.flatMap((n) => getAllRouteLeafs(n, opts)),
      ...(opts?.includeFailoverBranches && tree.failover
        ? getAllRouteLeafs(tree.failover, opts)
        : []),
    ];
  }

  if (tree.type === "leaf") {
    return [tree];
  }

  if (
    opts?.includeDeadLeafs &&
    (tree.type === "dead-leaf" || tree.type === "dead-branch")
  ) {
    return [tree];
  }

  return [];
}
