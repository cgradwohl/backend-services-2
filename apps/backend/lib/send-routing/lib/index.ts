export * from "./get-all-route-leafs";
export * from "./get-parent-route-node";
export * from "./get-parent-route-node-address";
export * from "./get-route-node";
export * from "./routing-strategy-store-helpers";
