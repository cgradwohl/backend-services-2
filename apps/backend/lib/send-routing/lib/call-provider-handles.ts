import { TokensByProvider } from "~/lib/token-storage";
import providers from "~/providers";
import { IProviderConfiguration } from "~/send/types";
import { JSONObject } from "~/types.api";

export const callProviderHandles = ({
  providerConfig,
  providerKey,
  data,
  profile,
  tokensByProvider,
}: {
  providerKey: string;
  providerConfig: IProviderConfiguration;
  data?: JSONObject;
  profile?: JSONObject;
  tokensByProvider?: TokensByProvider;
}): Promise<boolean | string | Error> | boolean | Error => {
  return (
    providers[providerKey]?.handles({
      config: providerConfig,
      data,
      profile,
      tokensByProvider,
    }) ?? false
  );
};
