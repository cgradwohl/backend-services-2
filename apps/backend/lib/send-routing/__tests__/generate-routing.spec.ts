import { applyTemplateChannelId, generateRouting } from "../generate-routing";
import * as callProviderHandlesModule from "../lib/call-provider-handles";
import { RouteLeaf, RoutingStrategy } from "../types";

describe("generate routing", () => {
  beforeEach(jest.clearAllMocks);

  it("should generate a valid tree from an all method routing config", async () => {
    expect.assertions(1);
    const strategy: RoutingStrategy = {
      routing: {
        method: "all",
        channels: ["sms", "email"],
      },
      channels: {},
      providers: {},
    };

    const routing = await generateRouting({
      strategy,
      providerConfigs,
      params: {
        data: {},
        profile: {
          phone_number: "1234567890",
          email: "drew@ycourier.com",
        },
      },
    });

    expect(routing).toMatchSnapshot();
  });

  it("should generate a valid tree from an single method routing config", async () => {
    expect.assertions(1);
    const strategy: RoutingStrategy = {
      routing: {
        method: "single",
        channels: ["sms", "email"],
      },
      channels: {},
      providers: {},
    };

    const routing = await generateRouting({
      strategy,
      providerConfigs,
      params: {
        data: {},
        profile: {
          phone_number: "1234567890",
          email: "drew@ycourier.com",
        },
      },
    });

    expect(routing).toMatchSnapshot();
  });

  it("should generate dead leafs for providers that could not be selected", async () => {
    expect.assertions(1);
    const strategy: RoutingStrategy = {
      routing: {
        method: "single",
        channels: ["sms", "email"],
      },
      channels: {},
      providers: {},
    };

    const routing = await generateRouting({
      strategy,
      providerConfigs,
      params: {
        data: {},
        profile: {
          email: "drew@ycourier.com",
        },
      },
    });

    expect(routing).toMatchSnapshot();
  });

  it("should produce a complex tree where a channel sends to more than one provider in a single call", async () => {
    expect.assertions(1);
    const strategy: RoutingStrategy = {
      routing: {
        method: "single",
        channels: ["push", "sms"],
      },
      channels: {
        push: {
          routing_method: "all",
          providers: ["apn", "firebase-fcm"],
        },
      },
      providers: {},
    };

    const routing = await generateRouting({
      strategy,
      providerConfigs,
      params: {
        data: {},
        profile: {
          phone_number: "1234567890",
          email: "drew@ycourier.com",
          apn: { token: "foo" },
          firebaseToken: "bar",
        },
      },
    });

    // The resulting tree should have both of these providers selected alongside each other.
    // It should also have an sms failover branch.
    expect(routing).toMatchSnapshot();
  });

  it("should evaluate configured channel conditionals and exclude the email channel as a result", async () => {
    expect.assertions(1);
    const strategy: RoutingStrategy = {
      routing: {
        method: "all",
        channels: ["sms", "email"],
      },
      channels: {
        sms: {
          if: "data.locale === 'en-US'",
        },
        email: {
          if: "data.locale === 'eu-FR'",
        },
      },
      providers: {},
    };

    const routing = await generateRouting({
      strategy,
      providerConfigs,
      params: {
        data: { locale: "en-US" },
        profile: { phone_number: "1234567890", email: "drew@ycourier.com" },
      },
    });

    expect(routing).toMatchSnapshot();
  });

  it("should evaluate provider conditionals and exclude mailjet as a result", async () => {
    expect.assertions(1);
    const strategy: RoutingStrategy = {
      routing: {
        method: "all",
        channels: ["sms", "email"],
      },
      channels: {
        sms: {
          providers: ["twilio"],
        },
        email: {
          providers: ["mailjet"],
        },
      },
      providers: {
        twilio: {
          if: "data.locale === 'en-US'",
        },
        mailjet: {
          if: "data.locale === 'eu-FR'",
        },
      },
    };

    const routing = await generateRouting({
      strategy,
      providerConfigs,
      params: {
        data: { locale: "en-US" },
        profile: { phone_number: "1234567890", email: "drew@courier.com" },
      },
    });

    expect(routing).toMatchSnapshot();
  });

  it("should allow providers to be channels", async () => {
    expect.assertions(1);
    const strategy: RoutingStrategy = {
      routing: {
        method: "all",
        channels: ["mailjet"],
      },
      channels: {},
      providers: {},
    };

    const routing = await generateRouting({
      strategy,
      providerConfigs,
      params: {
        data: {},
        profile: { email: "drew@courier.com" },
      },
    });

    expect(routing).toMatchSnapshot();
  });

  it("should handle misspelled channels and providers without puking", async () => {
    expect.assertions(1);
    const strategy: RoutingStrategy = {
      routing: {
        method: "all",
        channels: ["mailet", "eml"],
      },
      channels: {},
      providers: {},
    };

    const routing = await generateRouting({
      strategy,
      providerConfigs,
      params: {
        data: {},
        profile: { email: "drew@courier.com" },
      },
    });

    expect(routing).toMatchSnapshot();
  });

  it("should call provider handles function and use the result", async () => {
    expect.assertions(3);
    const strategy: RoutingStrategy = {
      routing: {
        method: "all",
        channels: ["vonage"],
      },
      channels: {},
      providers: {},
    };
    const opts = {
      strategy,
      providerConfigs,
      params: {
        data: {},
        profile: {},
      },
    };

    const mockCallProviderHandles = jest.spyOn(
      callProviderHandlesModule,
      "callProviderHandles"
    );

    // MISSING_PROVIDER_SUPPORT case
    mockCallProviderHandles.mockReturnValue(Promise.resolve(false));
    const routing = await generateRouting(opts);
    expect(routing).toMatchSnapshot();

    // PROVIDER_ERROR case
    mockCallProviderHandles.mockReturnValue(Promise.reject(new Error()));
    const routing2 = await generateRouting(opts);
    expect(routing2).toMatchSnapshot();

    // Success case
    mockCallProviderHandles.mockResolvedValue(Promise.resolve(true));
    const routing3 = await generateRouting(opts);
    expect(routing3).toMatchSnapshot();

    mockCallProviderHandles.mockRestore();
  });

  it("should pass tokens to provider handles functions", async () => {
    expect.assertions(1);
    const strategy: RoutingStrategy = {
      routing: {
        method: "all",
        channels: ["apn"],
      },
      channels: {},
      providers: {},
    };

    const apnTrees = await Promise.all([
      generateRouting({
        strategy,
        providerConfigs,
        params: {
          data: {},
          profile: {},
        },
      }),
      generateRouting({
        strategy,
        providerConfigs,
        tokens: { apn: [{ token: "hello!" } as any] },
        params: {
          data: {},
          profile: {},
        },
      }),
    ]);

    expect(apnTrees).toMatchSnapshot();
  });

  it("should generate a tree with recursive routing strategies", async () => {
    expect.assertions(1);
    const strategy: RoutingStrategy = {
      routing: {
        method: "all",
        channels: [
          {
            method: "single",
            channels: ["sms", "push"],
          },
          {
            method: "single",
            channels: ["email", "slack"],
          },
        ],
      },
      channels: {},
      providers: {},
    };

    const routing = await generateRouting({
      strategy,
      providerConfigs,
      params: {
        data: {},
        profile: {
          email: "drew@courier.com",
          phone_number: "1234567890",
          apn: {
            token: "YOUR TOKEN",
          },
        },
      },
    });

    expect(routing).toMatchSnapshot();
  });

  describe("applyLegacyTemplateChannelId", () => {
    const leaf: RouteLeaf = {
      channel: "email",
      providerConfigurationId: "b46abb6b-e99e-473b-927a-7d9843ec0442",
      provider: "mailjet",
      taxonomy: "email:mailjet",
      providerFailoverIndex: 1,
      type: "leaf",
      address: [],
    };

    it("should leave routing summary alone if no notification", () => {
      const result = applyTemplateChannelId(leaf);
      expect(result).toEqual(leaf);
    });

    it("should add relevant information if notification", () => {
      const result = applyTemplateChannelId(leaf, notification);
      expect(result.templateChannelId).toEqual(
        "6721af5a-7d56-4906-a021-6f30fe84253a"
      );
    });
  });
});

const providerConfigs: any = [
  {
    json: {
      domain: "sandboxb1d4dd0eb044497987e3c14f3c7c4db6.mailgun.org",
      apiKey: "4cea38609bdedcd992c18fd595a55a6d-7b8c9ba8-91e9abcd",
      fromAddress: "cmoney@example.com",
      provider: "apn",
    },
    id: "f2661ebd-dd91-417c-8b37-a9870fd0cbf4",
  },
  {
    json: {
      domain: "sandboxb1d4dd0eb044497987e3c14f3c7c4db6.mailgun.org",
      apiKey: "4cea38609bdedcd992c18fd595a55a6d-7b8c9ba8-91e9abcd",
      fromAddress: "cmoney@example.com",
      provider: "firebase-fcm",
    },
    id: "fsdkjhf-fdsf-sdf-sdfs-sdffsdf",
  },
  {
    json: {
      apiKey: "SG.gf.df",
      fromAddress: "chris@courier.com",
      checkDeliveryStatus: false,
      provider: "airship",
    },
    id: "a0e317ae-b30b-482f-a503-5e772e234f65",
  },
  {
    json: {
      accountSid: "wetwetgs",
      authToken: "tewthfdadfdf",
      messagingServiceSid: "greswetwtre",
      provider: "twilio",
    },
    id: "7593c6b6-f241-47f0-85b4-b5386cd60086",
  },
  {
    json: {
      accountSid: "wetwetgs",
      authToken: "tewthfdadfdf",
      messagingServiceSid: "greswetwtre",
      provider: "vonage",
    },
    id: "7593c6b6-f241-47f0-85b4-vonage",
  },
  {
    json: {
      accountSid: "wretewrtwq",
      authToken: "wtewetsgfdgssdgf",
      messagingServiceSid: "sdfgsdfgfdsdssdfg",
      provider: "mailjet",
    },
    id: "7593c6b6-f241-47f0-85b4-b5386cd600er6",
  },
  {
    json: {
      accountSid: "GSGDSewrewrewr",
      authToken: "gdfsgferger345sfa",
      messagingServiceSid: "sdfsggfdsdfsgfsdfger",
      provider: "sendgrid",
    },
    id: "2321",
  },
  {
    json: {
      accountSid: "gsdfds4353454",
      authToken: "324323423423",
      messagingServiceSid: "adsfasadfaswewr",
      provider: "slack",
    },
    id: "slack2e1231",
  },
];

const notification: any = {
  json: {
    channels: {
      always: [
        {
          channel: "email",
          channelLabel: "Email",
          provider: "mailjet",
          selected: true,
          key: "mailjet",
          configurationId: "b46abb6b-e99e-473b-927a-7d9843ec0442",
          config: {
            email: {
              emailSubject: "My Test Subject {name}",
              emailTemplateConfig: {
                templateName: "none",
                topBarColor: "#9121C2",
              },
              renderPlainText: false,
            },
            locales: {},
          },
          conditional: {
            filters: [
              {
                source: "data",
                property: "do_email",
                value: "true",
                operator: "EQUALS",
                id: "af4fc93d-4321-4737-9556-e8954d33a9ad",
              },
            ],
            logicalOperator: "and",
            behavior: "hide",
          },
          blockIds: ["92e08808-8883-4034-be2f-d372df5db3c1"],
          id: "6721af5a-7d56-4906-a021-6f30fe84253a",
          providers: [
            {
              key: "mailjet",
              configurationId: "b46abb6b-e99e-473b-927a-7d9843ec0442",
              config: {},
              conditional: {
                filters: [
                  {
                    source: "data",
                    property: "dont",
                    value: "true",
                    operator: "EQUALS",
                    id: "278a9468-8159-4518-8554-8212d7e3c9fd",
                  },
                ],
                logicalOperator: "and",
                behavior: "hide",
              },
            },
          ],
          taxonomy: "email:*",
          disabled: false,
          label: "Email",
        },
      ],
    },
  },
};
