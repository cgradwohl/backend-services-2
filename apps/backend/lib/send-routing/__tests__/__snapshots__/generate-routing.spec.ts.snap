// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generate routing should allow providers to be channels 1`] = `
Object {
  "address": Array [],
  "nodes": Array [
    Object {
      "address": Array [
        0,
      ],
      "failover": undefined,
      "nodes": Array [
        Object {
          "address": Array [
            0,
            0,
          ],
          "channel": "mailjet",
          "provider": "mailjet",
          "providerConfigurationId": "7593c6b6-f241-47f0-85b4-b5386cd600er6",
          "providerFailoverIndex": 1,
          "taxonomy": "email:mailjet",
          "type": "leaf",
        },
      ],
      "type": "branch",
    },
  ],
  "type": "branch",
}
`;

exports[`generate routing should call provider handles function and use the result 1`] = `
Object {
  "address": Array [],
  "nodes": Array [
    Object {
      "address": Array [
        0,
      ],
      "nodes": Array [
        Object {
          "address": Array [
            0,
            0,
          ],
          "channel": "vonage",
          "failureReason": "Information required by the provider was not included.",
          "failureType": "MISSING_PROVIDER_SUPPORT",
          "provider": "vonage",
          "type": "dead-leaf",
        },
      ],
      "type": "branch",
    },
  ],
  "type": "branch",
}
`;

exports[`generate routing should call provider handles function and use the result 2`] = `
Object {
  "address": Array [],
  "nodes": Array [
    Object {
      "address": Array [
        0,
      ],
      "nodes": Array [
        Object {
          "address": Array [
            0,
            0,
          ],
          "channel": "vonage",
          "failureReason": "Provider returned an error when checking its ability to handle the notification",
          "failureType": "PROVIDER_ERROR",
          "provider": "vonage",
          "type": "dead-leaf",
        },
      ],
      "type": "branch",
    },
  ],
  "type": "branch",
}
`;

exports[`generate routing should call provider handles function and use the result 3`] = `
Object {
  "address": Array [],
  "nodes": Array [
    Object {
      "address": Array [
        0,
      ],
      "failover": undefined,
      "nodes": Array [
        Object {
          "address": Array [
            0,
            0,
          ],
          "channel": "vonage",
          "provider": "vonage",
          "providerConfigurationId": "7593c6b6-f241-47f0-85b4-vonage",
          "providerFailoverIndex": 1,
          "taxonomy": "direct_message:sms:vonage",
          "type": "leaf",
        },
      ],
      "type": "branch",
    },
  ],
  "type": "branch",
}
`;

exports[`generate routing should evaluate configured channel conditionals and exclude the email channel as a result 1`] = `
Object {
  "address": Array [],
  "nodes": Array [
    Object {
      "address": Array [
        0,
      ],
      "failover": Object {
        "address": Array [
          0,
          "failover",
        ],
        "failover": undefined,
        "nodes": Array [
          Object {
            "address": Array [
              0,
              "failover",
              0,
            ],
            "channel": "sms",
            "provider": "vonage",
            "providerConfigurationId": "7593c6b6-f241-47f0-85b4-vonage",
            "providerFailoverIndex": 2,
            "taxonomy": "direct_message:sms:vonage",
            "type": "leaf",
          },
        ],
        "type": "branch",
      },
      "nodes": Array [
        Object {
          "address": Array [
            0,
            0,
          ],
          "channel": "sms",
          "provider": "twilio",
          "providerConfigurationId": "7593c6b6-f241-47f0-85b4-b5386cd60086",
          "providerFailoverIndex": 1,
          "taxonomy": "direct_message:sms:twilio",
          "type": "leaf",
        },
      ],
      "type": "branch",
    },
    Object {
      "address": Array [
        1,
      ],
      "channel": "email",
      "failureReason": "Channel conditional failed",
      "failureType": "FILTERED",
      "type": "dead-branch",
    },
  ],
  "type": "branch",
}
`;

exports[`generate routing should evaluate provider conditionals and exclude mailjet as a result 1`] = `
Object {
  "address": Array [],
  "nodes": Array [
    Object {
      "address": Array [
        0,
      ],
      "failover": undefined,
      "nodes": Array [
        Object {
          "address": Array [
            0,
            0,
          ],
          "channel": "sms",
          "provider": "twilio",
          "providerConfigurationId": "7593c6b6-f241-47f0-85b4-b5386cd60086",
          "providerFailoverIndex": 1,
          "taxonomy": "direct_message:sms:twilio",
          "type": "leaf",
        },
      ],
      "type": "branch",
    },
    Object {
      "address": Array [
        1,
      ],
      "nodes": Array [
        Object {
          "address": Array [
            1,
            0,
          ],
          "channel": "email",
          "failureReason": "Provider conditional failed (message.providers.mailjet.if)",
          "failureType": "FILTERED",
          "provider": "mailjet",
          "type": "dead-leaf",
        },
      ],
      "type": "branch",
    },
  ],
  "type": "branch",
}
`;

exports[`generate routing should generate a tree with recursive routing strategies 1`] = `
Object {
  "address": Array [],
  "nodes": Array [
    Object {
      "address": Array [
        0,
      ],
      "failover": Object {
        "address": Array [
          0,
          "failover",
        ],
        "failover": undefined,
        "nodes": Array [
          Object {
            "address": Array [
              0,
              "failover",
              0,
            ],
            "failover": Object {
              "address": Array [
                0,
                "failover",
                0,
                "failover",
              ],
              "nodes": Array [
                Object {
                  "address": Array [
                    0,
                    "failover",
                    0,
                    "failover",
                    0,
                  ],
                  "channel": "push",
                  "failureReason": "Information required by the provider was not included.",
                  "failureType": "MISSING_PROVIDER_SUPPORT",
                  "provider": "firebase-fcm",
                  "type": "dead-leaf",
                },
                Object {
                  "address": Array [
                    0,
                    "failover",
                    0,
                    "failover",
                    1,
                  ],
                  "channel": "push",
                  "failureReason": "Information required by the provider was not included.",
                  "failureType": "MISSING_PROVIDER_SUPPORT",
                  "provider": "airship",
                  "type": "dead-leaf",
                },
              ],
              "type": "branch",
            },
            "nodes": Array [
              Object {
                "address": Array [
                  0,
                  "failover",
                  0,
                  0,
                ],
                "channel": "push",
                "provider": "apn",
                "providerConfigurationId": "f2661ebd-dd91-417c-8b37-a9870fd0cbf4",
                "providerFailoverIndex": 1,
                "taxonomy": "push:apn",
                "type": "leaf",
              },
            ],
            "type": "branch",
          },
        ],
        "type": "branch",
      },
      "nodes": Array [
        Object {
          "address": Array [
            0,
            0,
          ],
          "failover": Object {
            "address": Array [
              0,
              0,
              "failover",
            ],
            "failover": undefined,
            "nodes": Array [
              Object {
                "address": Array [
                  0,
                  0,
                  "failover",
                  0,
                ],
                "channel": "sms",
                "provider": "vonage",
                "providerConfigurationId": "7593c6b6-f241-47f0-85b4-vonage",
                "providerFailoverIndex": 2,
                "taxonomy": "direct_message:sms:vonage",
                "type": "leaf",
              },
            ],
            "type": "branch",
          },
          "nodes": Array [
            Object {
              "address": Array [
                0,
                0,
                0,
              ],
              "channel": "sms",
              "provider": "twilio",
              "providerConfigurationId": "7593c6b6-f241-47f0-85b4-b5386cd60086",
              "providerFailoverIndex": 1,
              "taxonomy": "direct_message:sms:twilio",
              "type": "leaf",
            },
          ],
          "type": "branch",
        },
      ],
      "type": "branch",
    },
    Object {
      "address": Array [
        1,
      ],
      "failover": Object {
        "address": Array [
          1,
          "failover",
        ],
        "nodes": Array [
          Object {
            "address": Array [
              1,
              "failover",
              0,
            ],
            "nodes": Array [
              Object {
                "address": Array [
                  1,
                  "failover",
                  0,
                  0,
                ],
                "channel": "slack",
                "failureReason": "Information required by the provider was not included.",
                "failureType": "MISSING_PROVIDER_SUPPORT",
                "provider": "slack",
                "type": "dead-leaf",
              },
            ],
            "type": "branch",
          },
        ],
        "type": "branch",
      },
      "nodes": Array [
        Object {
          "address": Array [
            1,
            0,
          ],
          "failover": Object {
            "address": Array [
              1,
              0,
              "failover",
            ],
            "failover": undefined,
            "nodes": Array [
              Object {
                "address": Array [
                  1,
                  0,
                  "failover",
                  0,
                ],
                "channel": "email",
                "provider": "sendgrid",
                "providerConfigurationId": "2321",
                "providerFailoverIndex": 2,
                "taxonomy": "email:sendgrid",
                "type": "leaf",
              },
            ],
            "type": "branch",
          },
          "nodes": Array [
            Object {
              "address": Array [
                1,
                0,
                0,
              ],
              "channel": "email",
              "provider": "mailjet",
              "providerConfigurationId": "7593c6b6-f241-47f0-85b4-b5386cd600er6",
              "providerFailoverIndex": 1,
              "taxonomy": "email:mailjet",
              "type": "leaf",
            },
          ],
          "type": "branch",
        },
      ],
      "type": "branch",
    },
  ],
  "type": "branch",
}
`;

exports[`generate routing should generate a valid tree from an all method routing config 1`] = `
Object {
  "address": Array [],
  "nodes": Array [
    Object {
      "address": Array [
        0,
      ],
      "failover": Object {
        "address": Array [
          0,
          "failover",
        ],
        "failover": undefined,
        "nodes": Array [
          Object {
            "address": Array [
              0,
              "failover",
              0,
            ],
            "channel": "sms",
            "provider": "vonage",
            "providerConfigurationId": "7593c6b6-f241-47f0-85b4-vonage",
            "providerFailoverIndex": 2,
            "taxonomy": "direct_message:sms:vonage",
            "type": "leaf",
          },
        ],
        "type": "branch",
      },
      "nodes": Array [
        Object {
          "address": Array [
            0,
            0,
          ],
          "channel": "sms",
          "provider": "twilio",
          "providerConfigurationId": "7593c6b6-f241-47f0-85b4-b5386cd60086",
          "providerFailoverIndex": 1,
          "taxonomy": "direct_message:sms:twilio",
          "type": "leaf",
        },
      ],
      "type": "branch",
    },
    Object {
      "address": Array [
        1,
      ],
      "failover": Object {
        "address": Array [
          1,
          "failover",
        ],
        "failover": undefined,
        "nodes": Array [
          Object {
            "address": Array [
              1,
              "failover",
              0,
            ],
            "channel": "email",
            "provider": "sendgrid",
            "providerConfigurationId": "2321",
            "providerFailoverIndex": 2,
            "taxonomy": "email:sendgrid",
            "type": "leaf",
          },
        ],
        "type": "branch",
      },
      "nodes": Array [
        Object {
          "address": Array [
            1,
            0,
          ],
          "channel": "email",
          "provider": "mailjet",
          "providerConfigurationId": "7593c6b6-f241-47f0-85b4-b5386cd600er6",
          "providerFailoverIndex": 1,
          "taxonomy": "email:mailjet",
          "type": "leaf",
        },
      ],
      "type": "branch",
    },
  ],
  "type": "branch",
}
`;

exports[`generate routing should generate a valid tree from an single method routing config 1`] = `
Object {
  "address": Array [],
  "failover": Object {
    "address": Array [
      "failover",
    ],
    "failover": undefined,
    "nodes": Array [
      Object {
        "address": Array [
          "failover",
          0,
        ],
        "failover": Object {
          "address": Array [
            "failover",
            0,
            "failover",
          ],
          "failover": undefined,
          "nodes": Array [
            Object {
              "address": Array [
                "failover",
                0,
                "failover",
                0,
              ],
              "channel": "email",
              "provider": "sendgrid",
              "providerConfigurationId": "2321",
              "providerFailoverIndex": 2,
              "taxonomy": "email:sendgrid",
              "type": "leaf",
            },
          ],
          "type": "branch",
        },
        "nodes": Array [
          Object {
            "address": Array [
              "failover",
              0,
              0,
            ],
            "channel": "email",
            "provider": "mailjet",
            "providerConfigurationId": "7593c6b6-f241-47f0-85b4-b5386cd600er6",
            "providerFailoverIndex": 1,
            "taxonomy": "email:mailjet",
            "type": "leaf",
          },
        ],
        "type": "branch",
      },
    ],
    "type": "branch",
  },
  "nodes": Array [
    Object {
      "address": Array [
        0,
      ],
      "failover": Object {
        "address": Array [
          0,
          "failover",
        ],
        "failover": undefined,
        "nodes": Array [
          Object {
            "address": Array [
              0,
              "failover",
              0,
            ],
            "channel": "sms",
            "provider": "vonage",
            "providerConfigurationId": "7593c6b6-f241-47f0-85b4-vonage",
            "providerFailoverIndex": 2,
            "taxonomy": "direct_message:sms:vonage",
            "type": "leaf",
          },
        ],
        "type": "branch",
      },
      "nodes": Array [
        Object {
          "address": Array [
            0,
            0,
          ],
          "channel": "sms",
          "provider": "twilio",
          "providerConfigurationId": "7593c6b6-f241-47f0-85b4-b5386cd60086",
          "providerFailoverIndex": 1,
          "taxonomy": "direct_message:sms:twilio",
          "type": "leaf",
        },
      ],
      "type": "branch",
    },
  ],
  "type": "branch",
}
`;

exports[`generate routing should generate dead leafs for providers that could not be selected 1`] = `
Object {
  "address": Array [],
  "failover": undefined,
  "nodes": Array [
    Object {
      "address": Array [
        0,
      ],
      "nodes": Array [
        Object {
          "address": Array [
            0,
            0,
          ],
          "channel": "sms",
          "failureReason": "Information required by the provider was not included.",
          "failureType": "MISSING_PROVIDER_SUPPORT",
          "provider": "twilio",
          "type": "dead-leaf",
        },
        Object {
          "address": Array [
            0,
            1,
          ],
          "channel": "sms",
          "failureReason": "Information required by the provider was not included.",
          "failureType": "MISSING_PROVIDER_SUPPORT",
          "provider": "vonage",
          "type": "dead-leaf",
        },
      ],
      "type": "branch",
    },
    Object {
      "address": Array [
        1,
      ],
      "failover": Object {
        "address": Array [
          1,
          "failover",
        ],
        "failover": undefined,
        "nodes": Array [
          Object {
            "address": Array [
              1,
              "failover",
              0,
            ],
            "channel": "email",
            "provider": "sendgrid",
            "providerConfigurationId": "2321",
            "providerFailoverIndex": 2,
            "taxonomy": "email:sendgrid",
            "type": "leaf",
          },
        ],
        "type": "branch",
      },
      "nodes": Array [
        Object {
          "address": Array [
            1,
            0,
          ],
          "channel": "email",
          "provider": "mailjet",
          "providerConfigurationId": "7593c6b6-f241-47f0-85b4-b5386cd600er6",
          "providerFailoverIndex": 1,
          "taxonomy": "email:mailjet",
          "type": "leaf",
        },
      ],
      "type": "branch",
    },
  ],
  "type": "branch",
}
`;

exports[`generate routing should handle misspelled channels and providers without puking 1`] = `
Object {
  "address": Array [],
  "nodes": Array [
    Object {
      "address": Array [
        0,
      ],
      "channel": "mailet",
      "failureReason": "No remaining configured providers for channel or channel is invalid",
      "failureType": "NO_PROVIDERS",
      "type": "dead-branch",
    },
    Object {
      "address": Array [
        1,
      ],
      "channel": "eml",
      "failureReason": "No remaining configured providers for channel or channel is invalid",
      "failureType": "NO_PROVIDERS",
      "type": "dead-branch",
    },
  ],
  "type": "branch",
}
`;

exports[`generate routing should pass tokens to provider handles functions 1`] = `
Array [
  Object {
    "address": Array [],
    "nodes": Array [
      Object {
        "address": Array [
          0,
        ],
        "nodes": Array [
          Object {
            "address": Array [
              0,
              0,
            ],
            "channel": "apn",
            "failureReason": "Information required by the provider was not included.",
            "failureType": "MISSING_PROVIDER_SUPPORT",
            "provider": "apn",
            "type": "dead-leaf",
          },
        ],
        "type": "branch",
      },
    ],
    "type": "branch",
  },
  Object {
    "address": Array [],
    "nodes": Array [
      Object {
        "address": Array [
          0,
        ],
        "failover": undefined,
        "nodes": Array [
          Object {
            "address": Array [
              0,
              0,
            ],
            "channel": "apn",
            "provider": "apn",
            "providerConfigurationId": "f2661ebd-dd91-417c-8b37-a9870fd0cbf4",
            "providerFailoverIndex": 1,
            "taxonomy": "push:apn",
            "type": "leaf",
          },
        ],
        "type": "branch",
      },
    ],
    "type": "branch",
  },
]
`;

exports[`generate routing should produce a complex tree where a channel sends to more than one provider in a single call 1`] = `
Object {
  "address": Array [],
  "failover": Object {
    "address": Array [
      "failover",
    ],
    "failover": undefined,
    "nodes": Array [
      Object {
        "address": Array [
          "failover",
          0,
        ],
        "failover": Object {
          "address": Array [
            "failover",
            0,
            "failover",
          ],
          "failover": undefined,
          "nodes": Array [
            Object {
              "address": Array [
                "failover",
                0,
                "failover",
                0,
              ],
              "channel": "sms",
              "provider": "vonage",
              "providerConfigurationId": "7593c6b6-f241-47f0-85b4-vonage",
              "providerFailoverIndex": 2,
              "taxonomy": "direct_message:sms:vonage",
              "type": "leaf",
            },
          ],
          "type": "branch",
        },
        "nodes": Array [
          Object {
            "address": Array [
              "failover",
              0,
              0,
            ],
            "channel": "sms",
            "provider": "twilio",
            "providerConfigurationId": "7593c6b6-f241-47f0-85b4-b5386cd60086",
            "providerFailoverIndex": 1,
            "taxonomy": "direct_message:sms:twilio",
            "type": "leaf",
          },
        ],
        "type": "branch",
      },
    ],
    "type": "branch",
  },
  "nodes": Array [
    Object {
      "address": Array [
        0,
      ],
      "nodes": Array [
        Object {
          "address": Array [
            0,
            0,
          ],
          "channel": "push",
          "provider": "apn",
          "providerConfigurationId": "f2661ebd-dd91-417c-8b37-a9870fd0cbf4",
          "providerFailoverIndex": 1,
          "taxonomy": "push:apn",
          "type": "leaf",
        },
        Object {
          "address": Array [
            0,
            1,
          ],
          "channel": "push",
          "provider": "firebase-fcm",
          "providerConfigurationId": "fsdkjhf-fdsf-sdf-sdfs-sdffsdf",
          "providerFailoverIndex": 1,
          "taxonomy": "push:firebase-fcm",
          "type": "leaf",
        },
      ],
      "type": "branch",
    },
  ],
  "type": "branch",
}
`;
